### PR TITLE
Events: don't execute native stop(Immediate)Propagation from simulation

### PR DESCRIPTION
**Pull request by markelog**
In Firefox, called `stop(Immediate)Propagation` methods,
in capturing phase prevents receiving focus

See https://codepen.io/markelog/pen/pyBXOr.

/cc @bzbarsky

Probably not spec violation, but still pretty weird...

Fixes gh-3111

### CODE REVIEW

**Comment by markelog** [comment]
> capturing phase prevents receiving focus

Or maybe not "focus" exactly, but "caret"

File: src/event.js [hunk 1]
```diff
@@ -573,7 +574,7 @@ jQuery.Event.prototype = {
 
 		this.isPropagationStopped = returnTrue;
 
-		if ( e ) {
+		if ( e && this.executeOriginalPropagation ) {
```
Comment (markelog) [inline src/event.js]: I could have used `isSimulated` but choose not to, since such use wasn't intended for that property

File: test/unit/event.js [hunk 2]
```diff
@@ -2810,6 +2810,7 @@ QUnit.test( "originalEvent property for Chrome, Safari, Fx & Edge of simulated e
 	jQuery( "#qunit-fixture" ).append( html );
 
 	jQuery( "#donor-outer" ).on( "focusin", function( event ) {
+		console.log(1);
```
Comment (staabm) [inline test/unit/event.js]: Debug leftover

Comment (markelog) [inline test/unit/event.js]: Thank you, fixed

File: src/event/trigger.js [hunk 3]
```diff
@@ -150,26 +150,20 @@ jQuery.extend( jQuery.event, {
...
-				// For the compat branch though, guard for "click" and "submit"
-				// events is still used, but was moved to jQuery.event.stopPropagation function
-				// because `originalEvent` should point to the original event for the constancy
-				// with other events and for more focused logic
+				isSimulated: true,
+
+				// Support: Firefox <= 46
+				// In Firefox, called `stop(Immediate)Propagation` methods, in
+				// capturing phase, prevents receiving focus,
+				// so we stop only jQuery propagation not the native one instead
```
Comment (mgol) [inline src/event/trigger.js]: I'd tweak the grammar here:
```
// In Firefox, calling `stop(Immediate)Propagation` methods in
// the capture phase prevent receiving focus so we only stop
// the jQuery emulated propagation and not the native one.
```

File: test/unit/event.js [hunk 4]
```diff
@@ -2821,6 +2821,46 @@ QUnit.test( "originalEvent property for Chrome, Safari, Fx & Edge of simulated e
...
 } );
 
+QUnit.test(
+	"navite stop(Immediate)Propagation methods should not be called from simulated event",
+	function( assert ) {
+		var userAgent = window.navigator.userAgent;
+
+		if ( !( /firefox/i.test( userAgent ) || /safari/i.test( userAgent ) ) ) {
+			assert.expect( 1 );
+			assert.ok( true, "Assertions should run only in Chrome, Safari, Fx & Edge" );
```
Comment (mgol) [inline test/unit/event.js]: Why is that necessary?

Also, it might be simpler to say where we don't want to run them. :) It seems only in IE?

**Comment by mgol** [comment]
@markelog Please mention gh-3111 in the PR description so that it's easy to go to the issue.

File: test/unit/event.js [hunk 5]
```diff
@@ -2821,6 +2821,46 @@ QUnit.test( "originalEvent property for Chrome, Safari, Fx & Edge of simulated e
...
+			} )
+			.on( "focusin", function( event ) {
+				checker.immediate = sinon.stub( event.originalEvent, "stopImmediatePropagation" );
+				event.stopImmediatePropagation();
+			} );
+
+
+		jQuery( "#donor-input" ).trigger( "focus" );
+		assert.strictEqual( checker.simple.called, false );
+		assert.strictEqual( checker.immediate.called, false );
```
Comment (mgol) [inline test/unit/event.js]: That's checking our implementation but not the reported issue. Could we add a test that makes sure the input actually received focus?

File: src/event/trigger.js (hunk 3 above)
Comment (markelog) [inline src/event/trigger.js]: mm, not sure why current version doesn't suffice

File: test/unit/event.js (hunk 4 above)
Comment (markelog) [inline test/unit/event.js]: See test above this one

File: test/unit/event.js (hunk 5 above)
Comment (markelog) [inline test/unit/event.js]: Only with e2e tests

**Comment by markelog** [comment]
> @markelog Please mention gh-3111 in the PR description so that it's easy to go to the issue.

I did in commit message, github already handles did the referencing

File: test/unit/event.js (hunk 5 above)
Comment (markelog) [inline test/unit/event.js]: btw, it actually tests reported issue

File: src/event/trigger.js (hunk 3 above)
Comment (mgol) [inline src/event/trigger.js]: The current version has grammar errors.

File: test/unit/event.js (hunk 4 above)
Comment (mgol) [inline test/unit/event.js]: Ah, it's because it's emulated, ok.

File: test/unit/event.js [hunk 6]
```diff
@@ -2821,6 +2821,46 @@ QUnit.test( "originalEvent property for Chrome, Safari, Fx & Edge of simulated e
 	jQuery( "#donor-input" ).trigger( "focus" );
 } );
 
+QUnit.test(
+	"navite stop(Immediate)Propagation methods should not be called from simulated event",
```
Comment (mgol) [inline test/unit/event.js]: navite -> native

Comment (mgol) [inline test/unit/event.js]: "from a simulated event" or "from simulated events"

**Comment by mgol** [comment]
> I did in commit message, github already handles the referencing

To go from a PR to the issue you have to click on the commit tab, open the commit & only then click on the issue link. Having it in the description would make that easier (and that's the default when submitting a PR with the link to the issue already in the description).

File: test/unit/event.js (hunk 5 above)
Comment (mgol) [inline test/unit/event.js]: The issue was:

"The cursor is missing or misplaced in an input box with an focusin handler attached when returning false from the handler."

You could translate that to "Input doesn't receive focus if you invoke `preventDefault` or `stopPropagation` in the event handler, this doesn't test that, it only tests that the reason for this issue doesn't exist.

You'd test it only once we have e2e tests because focus-related tests are unreliable or why?

File: src/event/trigger.js (hunk 3 above)
Comment (markelog) [inline src/event/trigger.js]: What errors?

File: test/unit/event.js (hunk 5 above)
Comment (markelog) [inline test/unit/event.js]: > Input doesn't receive focus

It does receive focus, but not fully, it only represented with absent caret, it is happening because if `stopPropagation` or `stopImmediatePropagation`, **not** `preventDefault` is called on original, so here we test that it doesn't happen.

We can't actually trigger the real event, since it can be done only with e2e tests.

**Comment by markelog** [comment]
> To go from a PR to the issue you have to click...

There is reference to the issue, so that's that :), whereas discussion like this would be an example of waisted time

**Comment by bzbarsky** [comment]
@smaug---- for Mozilla event stuff.  That said, if focus is done by a listener and propagation is stopped... then focus won't happen.

File: src/event.js (hunk 1 above)
Comment (gibson042) [inline src/event.js]: In what sense is it different, though? The "Firefox <= 46" support comment could just move up here.

Comment (markelog) [inline src/event.js]: Comments would have to be duplicated in two places.

This is my preference to put it simply, i would like to be explicit on what this is and why it was introduced

**Comment by mgol** [comment]
> There is reference to the issue, so that's that :), whereas discussion like this would be an example of waisted time

It puzzles me you like to spend so much time arguing against a simple thing I asked you to do just because it makes it easier for me (& perhaps others) instead of just adding it. `¯\_(ツ)_/¯` I'll just end here.

**Comment by markelog** [comment]
@mgol you already posted reference to that issue with your **first** comment man, i really don't what else would you want, whereas following such rules which are not documented anywhere feels like nitpicking with the bad side. 

These things really do not matter, what matters is the bug which this pull trying to fix but instead of discussing that, you clobbering this issue with this nonsense - you could just post that link and notify me about your preference privately.

I really would just delete all these comments, feels like spam and waist of time to be honest to all involved.

> I'll just end here.

I would really like to not starting it at all.

File: test/unit/event.js (hunk 6 above)
Comment (markelog) [inline test/unit/event.js]: fixed

File: src/event/trigger.js (hunk 3 above)
Comment (markelog) [inline src/event/trigger.js]: To note: i will happily fix all potential errors, if you would point them out

**Comment by smaug----** [comment]
> That said, if focus is done by a listener and propagation is stopped... then focus won't happen.

This sounds like an old Gecko issue, IIRC there is a focus listener in editor code... Can't find the bug now though.

File: src/event.js (hunk 1 above)
Comment (gibson042) [inline src/event.js]: > Then comments would have to be duplicated in two places.

Where else would you have them?

File: src/event/trigger.js (hunk 3 above)
Comment (mgol) [inline src/event/trigger.js]: Methods (plural) prevent focus, not prevents, there are too many commas in the sentence, the action of calling the method prevents receiving focus and not the methods themselves, `only` sounds weird with `instead`, etc. You'll see that if you compare your comment with mine.

I've seen now that I made a mistake in my version as well: after changing `methods` to `calling` as the subject it's `prevents` again. Fixed.

File: src/event.js (hunk 1 above)
Comment (markelog) [inline src/event.js]: `stopImmediatePropagation` and `stopPropagation` methods

File: src/event/trigger.js [hunk 7]
```diff
@@ -150,26 +150,20 @@ jQuery.extend( jQuery.event, {
...
-				// But now, this "simulate" function is used only for events
-				// for which stopPropagation() is noop, so there is no need for that anymore.
-				//
-				// For the compat branch though, guard for "click" and "submit"
-				// events is still used, but was moved to jQuery.event.stopPropagation function
-				// because `originalEvent` should point to the original event for the constancy
-				// with other events and for more focused logic
+				isSimulated: true,
+
+				// Support: Firefox <= 46
```
Comment (markelog) [inline src/event/trigger.js]: @mgol changed, check it out

File: src/event.js (hunk 1 above)
Comment (gibson042) [inline src/event.js]: I see. But rereading the [`simulate` comments](https://github.com/jquery/jquery/blob/305f193aa57014dc7d8fa0739a3fefd47166cd44/src/event/trigger.js#L166-L167), it seems more like our folly than Mozilla's. I'd actually advocate for dropping the "Support" documentation here altogether in favor uniform simulated event treatment:
```js
stopPropagation: function() {
	var e = this.originalEvent;

	this.isPropagationStopped = returnTrue;

	// Stop propagation of original non-donor events
	if ( !this.isSimulated && e ) {
		e.stopPropagation();
	}
}
```
(and likewise for `stopImmediatePropagation`).

Comment (markelog) [inline src/event.js]: > But rereading the simulate comments,

If we would go to that route we would need the same thing for the `preventDefault`.

I think it was and it is reasonable, back then i checked with the spec of all donor events in case they had default actions, cancelable, etc; they all supposed to be noop - still holds true now too.

So it just waist of the space and judging by conversation of the parallel thread this is indeed mozilla's bug.

So these lines could be only "just in case" lines, without tests or use case, when/if mozilla people comes up with the fix

File: src/event/trigger.js (hunk 7 above)
Comment (mgol) [inline src/event/trigger.js]: Sounds good, thx.

File: src/event.js (hunk 1 above)
Comment (markelog) [inline src/event.js]: I can go either way though, the only requirement i would want to leave in is to save `originalEvent` property

File: test/unit/event.js (hunk 5 above)
Comment (mgol) [inline test/unit/event.js]: OK, got it, I thought it's being blurred in a way but I see I was wrong.

**Comment by markelog** [comment]
Changes are made

File: test/unit/event.js [hunk 8]
```diff
@@ -2790,6 +2790,81 @@ QUnit.test( "Donor event interference", function( assert ) {
...
+		assert.strictEqual( checker.prevent.called, false );
+
+		// We need to "off" it, since yes QUnit always update the fixtures
+		// but "focus" event listener is attached to document for focus(in | out)
+		// event and document doesn't get cleared obviously :)
+		outer.off( "focusin" );
+	}
+);
+
+QUnit.test(
```
Comment (dmethvin) [inline test/unit/event.js]: Why would you want to assert this? Isn't it an implementation detail? Just thinking about the case where Firefox implements the event and we would want to take out the code, this would fail.

Comment (markelog) [inline test/unit/event.js]: mm, i test here that property is present with `focus` event which is already implemented in firefox :). Here i wanted to make sure event has always consistent set of properties

**Comment by dmethvin** [comment]
Is this ready to land? I'm okay with it since `simulate` is only used for one thing and this change fixes the issue. We should get the  `.4` releases out this week.

**Comment by markelog** [comment]
Let say if there wouldn't be any objections for next 3-4 hours i will land this, if something comes up, we could always revert

**Comment by timmywil** [comment]
👍
